drv/psc-psu-update/src/main.rs

```diff
@@ -0,0 +1,262 @@
+
+                if revision == MWOCP68_FIRMWARE_REV {
+                    self.firmware_matches = Some(true);
+                    return false;
+                }
+
+                self.firmware_matches = Some(false);
+            }
+            Err(err) => {
+                ringbuf_entry!(Trace::FirmwareRevFailed(ndx, err));
+                return false;
+            }
+        }
+
+        if let (Some(started), Some(backoff)) =
+            (self.update_started, self.update_backoff)
+        {
+            if started + backoff > now {
+                ringbuf_entry!(Trace::BackingOff(ndx));
+                return false;
+            }
+        }
+
+        true
+    }
+
+    fn update_firmware(&mut self, dev: &Mwocp68, ndx: u8) {
+        ringbuf_entry!(Trace::AttemptingUpdate(ndx));
+        self.update_started = Some(Ticks::now());
+
+        self.update_backoff = match self.update_backoff {
+            Some(backoff) => Some(Ticks(backoff.0 * 2)),
+            None => Some(Ticks(60_000)),
+        };
+
+        let mut state = None;
+
+        loop {
+            match dev.update(state, MWOCP68_FIRMWARE_PAYLOAD) {
+                Err(err) => {
+                    //
+                    // We failed.  Record everything we can and leave.
+                    //
+                    ringbuf_entry!(Trace::UpdateFailed);
+                    ringbuf_entry!(Trace::UpdateFailedState(state));
+                    ringbuf_entry!(Trace::UpdateFailure(err));
+
+                    self.update_failure = Some((Ticks::now(), state, err));
+                    break;
+                }
+
+                Ok((UpdateState::UpdateSuccessful, _)) => {
+                    ringbuf_entry!(Trace::UpdateSucceeded);
+                    self.update_succeeded = Some(Ticks::now());
+                    break;
+                }
+
+                Ok((next, delay)) => {
+                    match next {
+                        UpdateState::WroteBlock { .. } => {
+                            ringbuf_entry!(Trace::WroteBlock);
+                        }
+                        _ => {
+                            ringbuf_entry!(Trace::UpdateState(next));
+                            ringbuf_entry!(Trace::UpdateDelay(delay));
+                        }
+                    }
+
+                    hl::sleep_for(delay);
+                    state = Some(next);
+                }
+            }
+        }
+    }
+}
+
+#[export_name = "main"]
+fn main() -> ! {
+    let i2c_task = I2C.get_task_id();
+
+    ringbuf_entry!(Trace::Start);
+
+    let mut psus = PSU.claim();
+
+    let devs: [Mwocp68; 6] = array_init::array_init(|ndx: usize| {
+        Mwocp68::new(&DEVICES[ndx](i2c_task), 0)
+    });
+
+    loop {
+        hl::sleep_for(TIMER_INTERVAL);
+
+        for (ndx, psu) in psus.iter_mut().enumerate() {
+            let dev = &devs[ndx];
+
+            if psu.update_should_be_attempted(dev, ndx as u8) {
+                psu.update_firmware(dev, ndx as u8);
```

Comment:
If a firmware update completes, but the rectifier doesn't report power-good (https://github.com/oxidecomputer/hubris/pull/1934/files#r1854779375), we should probably stop rather than rolling it out to all of the PSUs.  This would look like returning a status code from `update_firmware`, and bailing out (into an infinite sleep?) under certain error conditions.

Reply:
Added logic here. This really (really) Shouldn't Happen, so if we see PG not set on a PSU we just successfully updated, we will wait until we do see PG become set.